[CI] SnapshotDisruptionIT testDisruptionOnSnapshotInitialization failed in 6.8

`org.elasticsearch.discovery.SnapshotDisruptionIT testDisruptionOnSnapshotInitialization` failed in https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+6.8+multijob-unix-compatibility/os=ubuntu-16.04&&immutable/299/console (scan https://gradle-enterprise.elastic.co/s/cmj4ig7w72la6)

The error was:

```
13:27:37 ERROR   49.7s J2  | SnapshotDisruptionIT.testDisruptionOnSnapshotInitialization <<< FAILURES!
13:27:37    > Throwable #1: SnapshotException[[test-repo:test-snap-2/0FtjxNKfRQatK_1HB_Ri9g] Snapshot could not be read]; nested: SnapshotMissingException[[test-repo:test-snap-2/0FtjxNKfRQatK_1HB_Ri9g] is missing]; nested: NoSuchFileException[/var/lib/jenkins/workspace/elastic+elasticsearch+6.8+multijob-unix-compatibility/os/ubuntu-16.04&&immutable/server/build/testrun/integTest/J2/temp/org.elasticsearch.discovery.SnapshotDisruptionIT_CCE6937203AE3890-001/tempDir-002/repos/pSFBIYthUy/snap-0FtjxNKfRQatK_1HB_Ri9g.dat];
13:27:37    > 	at __randomizedtesting.SeedInfo.seed([CCE6937203AE3890:E639B5A225B6CB4F]:0)
13:27:37    > 	at org.elasticsearch.snapshots.SnapshotsService.snapshots(SnapshotsService.java:212)
13:27:37    > 	at org.elasticsearch.action.admin.cluster.snapshots.get.TransportGetSnapshotsAction.masterOperation(TransportGetSnapshotsAction.java:136)
13:27:37    > 	at org.elasticsearch.action.admin.cluster.snapshots.get.TransportGetSnapshotsAction.masterOperation(TransportGetSnapshotsAction.java:55)
13:27:37    > 	at org.elasticsearch.action.support.master.TransportMasterNodeAction.masterOperation(TransportMasterNodeAction.java:124)
13:27:37    > 	at org.elasticsearch.action.support.master.TransportMasterNodeAction$AsyncSingleAction$2.doRun(TransportMasterNodeAction.java:211)
13:27:37    > 	at org.elasticsearch.common.util.concurrent.ThreadContext$ContextPreservingAbstractRunnable.doRun(ThreadContext.java:751)
13:27:37    > 	at org.elasticsearch.common.util.concurrent.AbstractRunnable.run(AbstractRunnable.java:37)
13:27:37    > 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
13:27:37    > 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
13:27:37    > 	at java.lang.Thread.run(Thread.java:748)
13:27:37    > Caused by: SnapshotMissingException[[test-repo:test-snap-2/0FtjxNKfRQatK_1HB_Ri9g] is missing]; nested: NoSuchFileException[/var/lib/jenkins/workspace/elastic+elasticsearch+6.8+multijob-unix-compatibility/os/ubuntu-16.04&&immutable/server/build/testrun/integTest/J2/temp/org.elasticsearch.discovery.SnapshotDisruptionIT_CCE6937203AE3890-001/tempDir-002/repos/pSFBIYthUy/snap-0FtjxNKfRQatK_1HB_Ri9g.dat];
13:27:37    > 	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getSnapshotInfo(BlobStoreRepository.java:558)
13:27:37    > 	at org.elasticsearch.snapshots.SnapshotsService.snapshots(SnapshotsService.java:206)
13:27:37    > 	... 9 more
13:27:37    > Caused by: java.nio.file.NoSuchFileException: /var/lib/jenkins/workspace/elastic+elasticsearch+6.8+multijob-unix-compatibility/os/ubuntu-16.04&&immutable/server/build/testrun/integTest/J2/temp/org.elasticsearch.discovery.SnapshotDisruptionIT_CCE6937203AE3890-001/tempDir-002/repos/pSFBIYthUy/snap-0FtjxNKfRQatK_1HB_Ri9g.dat
13:27:37    > 	at sun.nio.fs.UnixException.translateToIOException(UnixException.java:86)
13:27:37    > 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:102)
13:27:37    > 	at sun.nio.fs.UnixException.rethrowAsIOException(UnixException.java:107)
13:27:37    > 	at sun.nio.fs.UnixFileSystemProvider.newByteChannel(UnixFileSystemProvider.java:214)
13:27:37    > 	at java.nio.file.Files.newByteChannel(Files.java:361)
13:27:37    > 	at java.nio.file.Files.newByteChannel(Files.java:407)
13:27:37    > 	at java.nio.file.spi.FileSystemProvider.newInputStream(FileSystemProvider.java:384)
13:27:37    > 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newInputStream(FilterFileSystemProvider.java:192)
13:27:37    > 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newInputStream(FilterFileSystemProvider.java:192)
13:27:37    > 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newInputStream(FilterFileSystemProvider.java:192)
13:27:37    > 	at org.apache.lucene.mockfile.HandleTrackingFS.newInputStream(HandleTrackingFS.java:92)
13:27:37    > 	at org.apache.lucene.mockfile.FilterFileSystemProvider.newInputStream(FilterFileSystemProvider.java:192)
13:27:37    > 	at org.apache.lucene.mockfile.HandleTrackingFS.newInputStream(HandleTrackingFS.java:92)
13:27:37    > 	at java.nio.file.Files.newInputStream(Files.java:152)
13:27:37    > 	at org.elasticsearch.common.blobstore.fs.FsBlobContainer.readBlob(FsBlobContainer.java:120)
13:27:37    > 	at org.elasticsearch.repositories.blobstore.ChecksumBlobStoreFormat.readBlob(ChecksumBlobStoreFormat.java:101)
13:27:37    > 	at org.elasticsearch.repositories.blobstore.BlobStoreFormat.read(BlobStoreFormat.java:90)
13:27:37    > 	at org.elasticsearch.repositories.blobstore.BlobStoreRepository.getSnapshotInfo(BlobStoreRepository.java:556)
13:27:37    > 	... 10 more
```

This did not reproduce locally on the 6.8 branch using:

```
./gradlew ':server:integTest'   -Dtests.seed=CCE6937203AE3890   -Dtests.class=org.elasticsearch.discovery.SnapshotDisruptionIT   -Dtests.method="testDisruptionOnSnapshotInitialization"   -Dtests.security.manager=true   -Dtests.locale=ja-JP-u-ca-japanese-x-lvariant-JP   -Dtests.timezone=America/Santo_Domingo   -Dcompiler.java=12   -Druntime.java=8
```

--- a/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
+++ b/server/src/main/java/org/elasticsearch/snapshots/SnapshotsService.java
@@ -206,6 +206,9 @@ public class SnapshotsService extends AbstractLifecycleComponent implements Clus
                 if (ignoreUnavailable) {
                     logger.warn(() -> new ParameterizedMessage("failed to get snapshot [{}]", snapshotId), ex);
                 } else {
+                    if (ex instanceof SnapshotException) {
+                        throw ex;
+                    }
                     throw new SnapshotException(repositoryName, snapshotId, "Snapshot could not be read", ex);
                 }
             }

--- a/server/src/test/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java
+++ b/server/src/test/java/org/elasticsearch/snapshots/SnapshotStatusApisIT.java
@@ -21,19 +21,26 @@ package org.elasticsearch.snapshots;
 import org.elasticsearch.Version;
 import org.elasticsearch.action.ActionFuture;
 import org.elasticsearch.action.admin.cluster.snapshots.create.CreateSnapshotResponse;
+import org.elasticsearch.action.admin.cluster.snapshots.get.GetSnapshotsRequest;
+import org.elasticsearch.action.admin.cluster.snapshots.get.GetSnapshotsResponse;
 import org.elasticsearch.action.admin.cluster.snapshots.status.SnapshotStatus;
 import org.elasticsearch.action.admin.cluster.snapshots.status.SnapshotsStatusRequest;
 import org.elasticsearch.client.Client;
 import org.elasticsearch.cluster.SnapshotsInProgress;
 import org.elasticsearch.common.settings.Settings;
 import org.elasticsearch.common.unit.TimeValue;
+import org.elasticsearch.core.internal.io.IOUtils;
+import org.elasticsearch.repositories.blobstore.BlobStoreRepository;
 
+import java.io.IOException;
+import java.nio.file.Path;
 import java.util.List;
 import java.util.concurrent.TimeUnit;
 
 import static org.elasticsearch.test.hamcrest.ElasticsearchAssertions.assertAcked;
 import static org.hamcrest.Matchers.equalTo;
 import static org.hamcrest.Matchers.greaterThan;
+import static org.hamcrest.Matchers.instanceOf;
 
 public class SnapshotStatusApisIT extends AbstractSnapshotIntegTestCase {
 
@@ -109,5 +116,25 @@ public class SnapshotStatusApisIT extends AbstractSnapshotIntegTestCase {
 
         logger.info("--> wait for snapshot to finish");
         createSnapshotResponseActionFuture.actionGet();
+    }
+
+    public void testExceptionOnMissingSnapBlob() throws IOException {
+        disableRepoConsistencyCheck("This test intentionally corrupts the repository");
+
+        logger.info("--> creating repository");
+        final Path repoPath = randomRepoPath();
+        assertAcked(client().admin().cluster().preparePutRepository("test-repo").setType("fs").setSettings(
+            Settings.builder().put("location", repoPath).build()));
+
+        logger.info("--> snapshot");
+        final CreateSnapshotResponse response =
+            client().admin().cluster().prepareCreateSnapshot("test-repo", "test-snap").setWaitForCompletion(true).get();
+
+        logger.info("--> delete snap-${uuid}.dat file for this snapshot to simulate concurrent delete");
+        IOUtils.rm(repoPath.resolve(BlobStoreRepository.SNAPSHOT_PREFIX + response.getSnapshotInfo().snapshotId().getUUID() + ".dat"));
+
+        GetSnapshotsResponse snapshotsResponse = client().admin().cluster()
+            .getSnapshots(new GetSnapshotsRequest(new String[] {"test-repo"}, new String[] {"test-snap"})).actionGet();
+        assertThat(snapshotsResponse.getFailedResponses().get("test-repo"), instanceOf(SnapshotMissingException.class));
     }
 }

